# Exploratory Data Analysis

This notebook explores the cleaned BoardGameGeek dataset to better understand the available information before building a machine-learning model.

The analysis will examine:

- the claened dataset structure;
- remaining missing values;
- the distribution of `Rating Average`;
- numerical feature distributions;
- relationships between numerical features and the prdiction target;
- the `Mechanics` and `Domains` columns;
- important findings that may affect feature engineering and modling

## Load and Verify the Cleaned Dataset

The cleaned BoardGameGeek dataset is loaded from the `data/processed` folder.

Before beggining the exploratory analysis, the file path, dataset dimensions, column names, 
duplicate records, and prediction target is checked to confirm that the correct dataset has been loaded.


In [11]:
# Import Pandas for working with tabular data
import pandas as pd

# Import Path for creating file paths that work
# reliably across different OS-s
from pathlib import Path


# Find the folder from were notebook is running
current_folder = Path.cwd()

# Notebook is inside the "notebooks" folder,
# so its parent folder is the main project folder
project_root = current_folder.parent

# Create the complete path to the cleaned dataset
cleaned_data_path = (
    project_root
    / "data"
    / "processed"
    / "bgg_cleaned.csv"
)

print("Current folder:", current_folder)
print("Project root:", project_root)
print("Cleaned dataset path:", cleaned_data_path)
print("File exists:", cleaned_data_path.exists())

Current folder: c:\Users\Zombie\Desktop\vscode-projects\board-game-rating-predictor\notebooks
Project root: c:\Users\Zombie\Desktop\vscode-projects\board-game-rating-predictor
Cleaned dataset path: c:\Users\Zombie\Desktop\vscode-projects\board-game-rating-predictor\data\processed\bgg_cleaned.csv
File exists: True


In [12]:
# Load the cleaned CSV file into a Pandas DataFrame
df_eda = pd.read_csv(cleaned_data_path)

print("Dataset loaded successfully.")
print("Dataset shape:", df_eda.shape)

Dataset loaded successfully.
Dataset shape: (20342, 14)


In [13]:
# Display the first five rows of the cleaned dataset
df_eda.head()

,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
0,174430.0,Gloomhaven,2017.0,1.0,4.0,120.0,14.0,42055,8.79,1,3.86,68323.0,"Action Queue, Action Retrieval, Campaign / Bat...","Strategy Games, Thematic Games"
1,161936.0,Pandemic Legacy: Season 1,2015.0,2.0,4.0,60.0,13.0,41643,8.61,2,2.84,65294.0,"Action Points, Cooperative Game, Hand Manageme...","Strategy Games, Thematic Games"
2,224517.0,Brass: Birmingham,2018.0,2.0,4.0,120.0,14.0,19217,8.66,3,3.91,28785.0,"Hand Management, Income, Loans, Market, Networ...",Strategy Games
3,167791.0,Terraforming Mars,2016.0,1.0,5.0,120.0,12.0,64864,8.43,4,3.24,87099.0,"Card Drafting, Drafting, End Game Bonuses, Han...",Strategy Games
4,233078.0,Twilight Imperium: Fourth Edition,2017.0,3.0,6.0,480.0,14.0,13468,8.70,5,4.22,16831.0,"Action Drafting, Area Majority / Influence, Ar...","Strategy Games, Thematic Games"


In [ ]:
# Define the column names that I expect the cleaned dataset contains
expected_columns = [
    "ID",
    "Name", 
    "Year Published",
    "Min Players",
    "Max Players",
    "Play Time",
    "Min Age",
    "Users Rated",
    "Rating Average",
    "BGG Rank",
    "Complexity Average",
    "Owned Users",
    "Mechanics",
    "Domains"
]


# Check that if loaded dataset is matchng the expected cleaned dataset
print("Correct number of rows:", df_eda.shape[0] == 20342)
print("Correct number of columns:", df_eda.shape[1] == 14)

print(
    "Column names match:",
    df_eda.columns.tolist() == expected_columns
)

print(
    "Fully duplicated rows:",
    df_eda.duplicated().sum()
)

print(
    "Missing Rating Average values:",
    df_eda["Rating Average"].isnull().sum()
)

Correct number of rows: True
Correct number of columns: True
Column names match: True
Fully duplicated rows: 0
Missing Rating Average values: 0


## Review Remaining Missing Valeus

Some missing values remain in the cleaned dataset intentionally.

These values include information that was already absent in the original dataset and zero placeholders that were converted to proper missing values during data cleaning.

The number and percenage of missing values are reviewed before deciding how they should be handled during feature engineering and model preparation.


In [ ]:
# Count missing values is every column
missing_summary = df_eda.isnull().sum().to_frame(
    name="Missing Values"
)

# Calculate the percentage of the dataset
# that is missing in every column
# then divide each count by len(df_eda)
# then multiply by 100 - to changes to percentage
missing_summary["Missing Percentage"] = (
    missing_summary["Missing Values"]
    / len(df_eda)
    * 100 
).round(2)

# Keep only columns that contain mimimum one missing value
missing_summary = missing_summary[
    missing_summary["Missing Values"] > 0
]

# Sort missing percentage from largest > smallest
missing_summary = missing_summary.sort_values(
    by="Missing Percentage",
    ascending=False
)

missing_summary

,Missing Values,Missing Percentage
Min Age,1250,6.14
Play Time,556,2.73
Complexity Average,426,2.09
Year Published,185,0.91
Max Players,161,0.79
Min Players,46,0.23
Owned Users,22,0.11
ID,15,0.07


### Remaining Missing Values Result

The cleaned dataset contains remaining missing values in eight columns.

`Min Age` has the largest amunt of missing information, with 1,250 missing values, representing 6.14% of the dataset. The remaining columns each have less than 3% missing data.

Most of these missing values were intentionslly retained or created during data cleaning when invalid zero placeholders were converted to proper missing values.

The prediction target, `Rating Average`, contains no missing values.

No additional values will be filled or removed druing exploratory data analysis. The missing feature values will be handled later during feature engineering and model preprocessing to avoid making modelling decisions too erly.
